flight/Libraries/math/statistics.c

```diff
@@ -0,0 +1,352 @@
+/**
+ ******************************************************************************
+ * @addtogroup TauLabsMath Tau Labs math support libraries
+ * @{
+ * @addtogroup TauLabsStatistics Tau Labs statistics math support libraries
+ * @{
+ *
+ * @file       statistics.c
+ * @author     Kenn Sebesta, Copyright (C) 2015.
+ * @brief      Statistics support
+ *
+ *****************************************************************************/
+
+#include "statistics.h"
+
+#include <math.h>
+
+
+/**
+ * @brief bayes_filter Implements a bayesian testing under the effect of an
+ * actuator that can modify what is being sensed.
+ * Belief_H0 is the H0 hypothesis (also called the NULL HYPOTHESIS), e.g.
+ * the vehicle has landed.
+ * Belief_H1 is the H1 hypothesis (also called the NULL HYPOTHESIS), e.g.
+ * the vehicle is still flying.
+ *
+ * Inspired by Probablistic Robotics, Chap. 2.4, "Bayes Filters"
+ * @param belief_H0 The H0 hypothesis (also called the NULL HYPOTHESIS), e.g.
+ * the vehicle has landed.
+ * @param belief_H1 The H1 hypothesis (also called the NULL HYPOTHESIS), e.g.
+ * the vehicle is still flying.
+ * @param p_sense_condition_H0 The probability of sensing the H0 condition, given the real condition
+ * @param p_sense_condition_H1 The probability of sensing the H1 condition, given the real condition
+ * @param p_H0_Ux_H0 The probability of state H0 staying H0 after the control is applied
+ * @param p_H0_Ux_H1 The probability of state H1 becoming H0 after the control is applied
+ * @param p_H1_Ux_H0 The probability of state H0 becoming H1 after the control is applied
+ * @param p_H1_Ux_H1 The probability of state H1 staying H1 after the control is applied
+ */
+void bayes_filter(double *belief_H0, double *belief_H1,
```

Comment:
> really needs precision and it's hard to envision running it on a large data set at single precision with larger values.

Yeah, for sure! In fact, in my experience all the statistics functions should go to pure doubles. Single precision causes headaches because things look like they should work and then wind up not in practice because of simple design decisions like increasing the number of sample points. Bayesian hypothesis testing was one of those areas, because even five 9's of certainty is low (unless you have an excellent understanding of your PDF, which is unlikely and not particularly valuable in many scenarios).

I think it'd be cool if someone were to realize there's a neat application on the F1 and then use this code as the basis for a reduced precision version. If there were a specific, well-bounded question that we were looking to ask I bet there's a good chance we could to optimize all the math to fit. But I wrote this to be a generic solution so it's a little early for optimization. The risk of reducing precision is that people will use the lower precision in a knee-jerk fashion, not realizing the ramifications, and wind up having a miserable experience.